### ISSUE TITLE
Introduce expandable subset code snippets

### ISSUE THREAD

**Issue by parlough** [dart-lang/site-www#4908]
### Describe the problem

In many code snippets, we include code that isn't necessarily runnable on its own, such as code that belongs in a function. That allows us to focus on the important piece of the snippet. However, sometimes users expect to be able to run the snippet or try it out, but without a surrounding function or sometimes other code, it isn't possible. The resulting diagnostics or different behavior can be confusion.

### Expected fix

We should introduce a mechanism which allows us to focus on a certain portion of a snippet, but allow it to be expanded to be runnable. This would also make it possible to integrate with the dart analyzer highlighter and make every snippet runnable.

**Comment by MaryaBelanger** [comment]
Was just about to create a new issue, but found this! My issue was going to say, we could link in the snippets to the source (based on[ this comment ](https://github.com/dart-lang/site-www/issues/4938#issuecomment-1569873778)). 

I'm going to paste here what I was writing in the new issue, but lmk if you want to move it to it's own issue since it's a _slightly_ different idea:

---

Users aren't seeing that, e.g. variables are declared a certain way that makes the example possible, because those variables are declared outside of the `docregion` that we're trying to focus on. 

Also, we've had a couple instances of contributors editing code snippets when they come across these issues, which is _amazing_ and we want to encourage and make as easy as possible. But they understandably miss that there's also a code excerpt to update as well. 

It's not _that_ big a deal, but it'd be a smoother experience for consumers _and_ contributors of our documentation if that were obvious.

I'm imagining something like how GitHub Gists look in Medium articles:

![image](https://github.com/dart-lang/site-www/assets/111139605/da1a6ad0-cb6b-4520-8f1e-7e9cb25bc80e)

Could we add a link to the source file in the already existing margin of the snippets? 

![image](https://github.com/dart-lang/site-www/assets/111139605/561deecc-5558-4b30-ab6f-e63ee9b94c9c)

We already have a similar concept of "something in the margin" of some code snippets in practice:

![image](https://github.com/dart-lang/site-www/assets/111139605/48d7e5c0-1f76-48d4-a45c-2f0d007166da)

It'd be great if the link could:
- automatically populate just from the `<code-excerpt ...>` tag in the md
- link not only to the source code file, but the exact doc region

**Comment by MaryaBelanger** [comment]
Expanding is an interesting solution too. It would solve the problem for consumers of the docs, but I don't think it would solve the problem for contributors who are unaware of the source file they need to also update when editing snippets. 

Also, consider expanding a long source file. How much do we expand? The one page users have been having issues with lately, for example, [`pattern_types.dart`](https://github.com/dart-lang/site-www/blob/main/examples/language/lib/patterns/pattern_types.dart) is really long. If they're looking for `a` and `b` declared on [38-39](https://github.com/dart-lang/site-www/blob/main/examples/language/lib/patterns/pattern_types.dart#L38), they'd expand all the way to the top of the >250 line source file. 

These are just initial impressions, I'm not sure what the best way to solve this is. I also am not sure how integrating with the analyzer highlighter fits in, so that might be a deciding factor on choosing a solution.

**Comment by parlough** [comment]
As I have time, I'm working on a design doc for this work as it will involve solving a lot of related issues, so I'll focus on that instead. Once that is ready we can have a larger discussion.

At least with our current setup, I am cautious about linking to the source though. Currently our source `.dart` files make little sense, have a lot of extra unrelated structure, ignores, docregion comments, etc. to make things work. They are not true examples and the snippets only make sense in isolation. I'm afraid linking to it will lead to more confusion for docs consumers as they do not follow Dart rules. 

Whether part of expanding or linking (or both), examples would need to be made more complete, with more context, with less ignores, and more often runnable. Also many more examples would need to be in standalone files rather than one massive file.

In the mean time, we should likely make snippets as understandable as possible without referring to the source, as with these features or without, it's the goal they usually shouldn't need the rest to understand.

**Comment by MaryaBelanger** [comment]
Ok, definitely agree with all of your points. When I was imagining "expanding", I assumed they'd be expanding the entire `.dart` file that the snippet lives in, as is. So if that's not the case and we can somehow choose what's shown on expansion, that would be much much better than just a link back to the file. Or of course, like you said, just getting into the habit of writing cleaner files and standardizing how they're sectioned off would solve that problem too. 

> In the mean time, we should likely make snippets as understandable as possible without referring to the source, as with these features or without, it's the goal they usually shouldn't need the rest to understand.

Definitely. This recent handful of issues and fix-PRs for the snippets has taught me a very important lesson about code and context!

**Comment by parlough** [comment]
Yeah we for sure wouldn't want it to expand our current `.dart` files :P. I'm hoping the design solves those issues. I haven't yet decided whether examples should be their own file or if we specify the expandable region yet, but one will definitely be part of that plan. As you mentioned, it also depends on what I can figure out with the analyzer :) 

I'm hoping I can design it in a way that would leave most code updates in only `.dart` files rather than both locations as well, hopefully solving that concern.
